src/System.Security.Claims/src/System/Security/Claims/ClaimsIdentity.cs

```diff
@@ -0,0 +1,970 @@
+            }
+
+            Contract.EndContractBlock();
+
+            foreach (Claim claim in claims)
+            {
+                if (claim == null)
+                {
+                    continue;
+                }
+
+                if (object.ReferenceEquals(claim.Subject, this))
+                {
+                    _instanceClaims.Add(claim);
+                }
+                else
+                {
+                    _instanceClaims.Add(claim.Clone(this));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Attempts to remove a <see cref="Claim"/> the internal list. 
+        /// <summary/>
+        /// <param name="claim">the <see cref="Claim"/> to match.</param>
+        /// <remarks> It is possible that a <see cref="Claim"/> returned from <see cref="Claims"/> cannot be removed. This would be the case for 'External' claims that are provided by reference.
+        /// <para>object.ReferenceEquals is used to 'match'.</para>
+        /// </remarks>  
+        [SecurityCritical]
+        public virtual bool TryRemoveClaim(Claim claim)
+        {
+            if (claim == null)
+            {
+                return false;
+            }
+
+            bool removed = false;
+
+            for (int i = 0; i < _instanceClaims.Count; i++)
+            {
+                if (object.ReferenceEquals(_instanceClaims[i], claim))
+                {
+                    _instanceClaims.RemoveAt(i);
+                    removed = true;
+                    break;
+                }
+            }
+            return removed;
+        }
+
+        /// <summary>
+        /// Removes a <see cref="Claim"/> from the internal list. 
+        /// <summary/>
+        /// <param name="claim">the <see cref="Claim"/> to match.</param>
+        /// <remarks> It is possible that a <see cref="Claim"/> returned from <see cref="Claims"/> cannot be removed. This would be the case for 'External' claims that are provided by reference.
+        /// <para>object.ReferenceEquals is used to 'match'.</para>
+        /// </remarks>
+        /// <exception cref="InvalidOperationException">if 'claim' cannot be removed.</exception>
+        [SecurityCritical]
+        public virtual void RemoveClaim(Claim claim)
+        {
+            if (!TryRemoveClaim(claim))
+            {
+                throw new InvalidOperationException(
+                    string.Format(CultureInfo.InvariantCulture, 
+                                  SR.GetResourceString("InvalidOperation_ClaimCannotBeRemoved", "The Claim '{0}' was not able to be removed.  It is either not part of this Identity or it is a claim that is owned by the Principal that contains this Identity. For example, the Principal will own the claim when creating a GenericPrincipal with roles. The roles will be exposed through the Identity that is passed in the constructor, but not actually owned by the Identity.  Similar logic exists for a RolePrincipal."),
```

Comment:
As noted earlier, we should use the same mechanism for resources used elsewhere in corefx.  This can use ```SR.Format(SR.InvalidOperation_ClaimCannotBeRemoved, claim)```.